frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -531,9 +564,9 @@ export default ({
       }
 
       sbp('okTurtles.events/once', hash, async (contractID, message) => {
-        if (contractID === this.currentChatRoomId) {
-          const state = this.getSimulatedState(false)
-          await sbp('chelonia/private/in/processMessage', message, state)
+        // NOTE: while syncing the chatroom contract, we should ignore all the events
```

Comment:
I don't quite understand this `sbp('okTurtles.events/once', hash` event listener.

This function, `listenChatRoomActions`, will be called by `emitMessageEvent` in the `sideEffect`. `emitMessageEvent` could easily pass in `contractID` along with `hash`. I do not see a reason for immediately setting up an event listener here on `hash`, because it will simply be immediately called after the side effect in `handleEvent`: `sbp('okTurtles.events/emit', hash, contractID, message)`

What I'm saying is, I'm pretty sure you can remove `sbp('okTurtles.events/once', hash, async (contractID, message) => {`, and place the code inside of it into this function directly, right? (as long as you modify `emitMessageEvent` to pass along the `contractID` too)

Reply:
That's reasonable. The reasons why I made this way are because
1. I wanted to use `message` object in the function
2. As long as `listenChatRoomActions` is sync, there won't be any issues because `okTurtles.events/emit` is defined just like the following.
```
'okTurtles.events/emit': function (event: string, ...data: any) {
  for (const listener of sbp('okTurtles.data/get', listenKey(event)) || []) {
    listener(...data)
  }
}
```

Reply:
If I remove that event handler `sbp('okTurtles.events/once', hash)`, I am not sure how I can get the `message` object in `listenChatRoomActions` function.

Reply:
Oh! Good point! Can you add this comment above line 569?

```js
// we use this listener here so that we can get access to the GIMessage
// it will be called immediately after this function is called
```

Reply:
Sure.

Reply:
Yep! Of course, more informative is always better. :)

Although, that last part might have a typo:

>  as long as listenChatRoomActions function is sync

Did you mean "is syncing"? or "is synced"? Or what?

Reply:
`listenChatRoomActions is sync function`